client/components/signup/components/register/register.js

```diff
@@ -9,33 +12,58 @@ class Register extends PureComponent {
 		onSubmit: PropTypes.func.isRequired
 	}
 
-	usernameEl = null
-
 	state = {
 		username: '',
 		email: '',
-		password: ''
+		password: '',
+		showErrMsg: false,
+		errMsg: ''
 	}
 
+	usernameEl = null
+	errorIndex = 0
+
 	componentDidMount () {
 		this.usernameEl.focus()
 	}
 
+	hideErrMsg = () => {
+		this.setState({ showErrMsg: false })
+	}
+
 	handleGenericInput = evt => {
 		const tgt = evt.target
 		const value = tgt[tgt.type === 'checkbox' ? 'checked' : 'value']
 
 		this.setState({ [tgt.name]: value })
 	}
 
-	handleSubmit = evt => {
+	handleSubmit = async evt => {
 		evt.preventDefault()
 
 		const { username, email, password } = this.state
 
 		if (!username || !password) return
 
-		this.props.onSubmit(username, email, password)
+		const res = await this.props.onSubmit(username, email, password)
+
+		if (!res.success) {
+			this.setState({
+				showErrMsg: true,
+				errMsg: res.message || 'An unknown error occurred.'
+			}, () => {
+				this.errorIndex++
+
+				const thisIndex = this.errorIndex
+
+				sleep(3000)
+					.then(() => {
+						if (thisIndex !== this.errorIndex) return
```

Comment:
Same here with respect to adding braces after an `if` statement for increased clarity.

Reply:
I think (not sure) when I wrote this I had the idea in my mind that the functionality between login and register may diverge somewhat at some point, so an abstraction may not be beneficial. I'll take another look.

Reply:
If the functionality diverges then you can take that into account using a higher-order function (a function that takes in another function as a parameter). This will allow you to change only one place should you need to modify this common logic. However, if the result is not idiomatic then the current setup is the best approach.